src/troute-network/troute/DataAssimilation.py

```diff
@@ -115,4 +386,438 @@ def last_obs(self):
+    def __init__(self, data_assimilation_parameters, run_parameters, waterbody_parameters, network, da_run):
+        lastobs_df, da_parameter_dict = nnu.build_data_assimilation_lastobs(
+            data_assimilation_parameters
+        )
+    
+        # replace link ids with lake ids, for gages at waterbody outlets, 
+        # otherwise, gage data will not be assimilated at waterbody outlet
+        # segments.
+        if network.link_lake_crosswalk:
+            lastobs_df = _reindex_link_to_lake_id(lastobs_df, network.link_lake_crosswalk)
+            
+        self._last_obs_df = lastobs_df
+        self._da_parameter_dict = da_parameter_dict
+        # TODO: Add parameters here for interpolation length (14/59), QC threshold (1.0)
+        
+        #---------------------------------------------------------------------------
+        # Determine which DA features to turn on
+        #---------------------------------------------------------------------------
+
+        # isolate user-input parameters for streamflow data assimilation
+        streamflow_da_parameters = data_assimilation_parameters.get('streamflow_da', None)
+
+        # determine if user explictly requests streamflow DA
+        nudging = False
+        if streamflow_da_parameters:
+            nudging = streamflow_da_parameters.get('streamflow_nudging', False)
+
+        usgs_timeslices_folder = data_assimilation_parameters.get('usgs_timeslices_folder',None)
+        
+        # isolate user-input parameters for reservoir data assimilation
+        reservoir_da_parameters = data_assimilation_parameters.get('reservoir_da', None)
+
+        # check if user explictly requests USGS and/or USACE reservoir DA
+        usgs_persistence  = False
+        usace_persistence = False
+        if reservoir_da_parameters:
+            usgs_persistence  = reservoir_da_parameters.get('reservoir_persistence_usgs', False)
+            usace_persistence = reservoir_da_parameters.get('reservoir_persistence_usace', False)
+            param_file   = reservoir_da_parameters.get('gage_lakeID_crosswalk_file',None)
+        
+        # check if RFC-type reservoirs are set to true
+        rfc_params = waterbody_parameters.get('rfc')
+        if rfc_params:
+            rfc_forecast = rfc_params.get('reservoir_rfc_forecasts',False)
+            param_file = rfc_params.get('reservoir_parameter_file',None)
+        else:
+            rfc_forecast = False
+        
+        level_pool_params = waterbody_parameters.get('level_pool', defaultdict(list))
+                
+                
+        #--------------------------------------------------------------------------------
+        # Assemble USGS observation data for Streamflow DA or USGS Reservoir Persistence
+        #--------------------------------------------------------------------------------
+
+        # if user requested nudging or usgs_persistence and a specified a USGS TimeSlice directory, 
```

Comment:
slightly confusing comment "and a specified a USGS TimeSlice directory"

Reply:
dropped the 'a' before 'specified'. I think that makes it clearer?